packages/@aws-cdk/aws-iam/test/test.role.ts

```diff
@@ -309,45 +314,757 @@ export = {
+          statements: [new PolicyStatement({
+            actions: ['s3:*'],
+            resources: ['xyz'],
+          })],
+        }));
+
+        // THEN
+        expect(policyStack).to(haveResourceLike('AWS::IAM::Policy', {
+          PolicyDocument: {
+            Statement: [
+              {
+                Action: "s3:*",
+                Effect: "Allow",
+                Resource: "xyz",
+              },
+            ],
+          },
+          Roles: ["MyRole"],
+        }));
+
+        test.done();
+      },
+
+      // tslint:disable-next-line:max-line-length
+      'attaches policies from a different stack if imported by static ARN with default options and the ARN account == policy stack account'(test: Test) {
+        // GIVEN
+        const roleStack = new Stack();
+        const importedRole = Role.fromRoleArn(roleStack, 'ImportedRole',
+          'arn:aws:iam::123456789012:role/MyRole');
+
+        // WHEN
+        const policyStack = new Stack(undefined, undefined, { env: { account: '123456789012' }});
+        importedRole.attachInlinePolicy(new Policy(policyStack, 'MyPolicy', {
+          statements: [new PolicyStatement({
+            actions: ['s3:*'],
+            resources: ['xyz'],
+          })],
+        }));
+
+        // THEN
+        expect(policyStack).to(haveResourceLike('AWS::IAM::Policy', {
+          PolicyDocument: {
+            Statement: [
+              {
+                Action: "s3:*",
+                Effect: "Allow",
+                Resource: "xyz",
+              },
+            ],
+          },
+          Roles: ["MyRole"],
+        }));
+
+        test.done();
+      },
+
+      // tslint:disable-next-line:max-line-length
+      'ignores policies from a different stack if imported by static ARN with default options and the ARN account != policy stack account'(test: Test) {
+        // GIVEN
+        const roleStack = new Stack();
+        const importedRole = Role.fromRoleArn(roleStack, 'ImportedRole',
+          'arn:aws:iam::123456789012:role/MyRole');
+
+        // WHEN
+        const policyStack = new Stack(undefined, undefined, { env: { account: '012345678901' }});
+        importedRole.attachInlinePolicy(new Policy(policyStack, 'MyPolicy', {
+          statements: [new PolicyStatement({
+            actions: ['s3:*'],
+            resources: ['xyz'],
+          })],
+          roles: [new Role(policyStack, 'OtherRole', { assumedBy: new AnyPrincipal() })],
+        }));
+
+        // THEN
+        expect(policyStack).to(haveResourceLike('AWS::IAM::Policy', {
+          PolicyDocument: {
+            Statement: [
+              {
+                Action: "s3:*",
+                Effect: "Allow",
+                Resource: "xyz",
+              },
+            ],
+          },
+          Roles: [{ "Ref": "OtherRole30A5E499" }],
+        }));
+
+        test.done();
+      },
+
+      'adds statements to default policy if imported by dynamic ARN with default options'(test: Test) {
```

Comment:
I'm starting to get cross-eyed on these tests. We should factor out the setup and verification logic so that we can write more succinct statements, something along the lines of:

```ts
const physicalArn = 'arn:aws:12345:...';
const sameAccount = '12345';
const otherAccount = '6789';
const tokenValue = Lazy.stringValue({ ... });

addsStatement(physicalArn, { account: sameAccount }, { mutable: undefined });
addsPolicy(physicalArn, { account: sameAccount }, { mutable: undefined });

addsStatement(tokenValue, { account: sameAccount }, { mutable: undefined });
addsPolicy(tokenValue, { account: sameAccount }, { mutable: undefined });

doesntAddStatement(physicalArn, { account: otherAccount }, { mutable: undefined });
```

etc.

What do you think?

Reply:
I think we've reached the limits of NodeUnit here. I've migrated the IAM tests to Jest here: #3890 , and I'll base the new revision on that PR.